py-polars/src/lazyframe/visitor/expr_nodes.rs

```diff
@@ -571,8 +571,80 @@ pub(crate) fn into_py(py: Python<'_>, expr: &AExpr) -> PyResult<PyObject> {
                 FunctionExpr::ListExpr(_) => {
                     return Err(PyNotImplementedError::new_err("list expr"))
                 },
-                FunctionExpr::StringExpr(_) => {
-                    return Err(PyNotImplementedError::new_err("string expr"))
+                FunctionExpr::StringExpr(strfun) => match strfun {
+                    StringFunction::ConcatHorizontal {
+                        delimiter,
+                        ignore_nulls,
+                    } => ("concathorizontal", delimiter, ignore_nulls).to_object(py),
```

Comment:
One question I have here: we're effectively using a tuple with a string as a poor approximation of a tagged union type (AKA, the incoming enum type on the rust side).

This means that in Python it's not easily to introspect the set of values that we can receive here.

One alternative option (which is more typing) is to have an enum class here that is exposed. This could _either_ be a unit variant enum that replaces the use of the string "tag" with an enum value:

```rust
#[pyclass]
enum PyStringFunction {
  ConcatHorizontal,
  ConcatVertical,
  ...
}
...

match strfun {
   StringFunction::ConcatHorizontal { delimiter, ignore_nulls } => (PyStringFunction, delimiter, ignore_nulls).to_object(py),
   ...
}
```

Or, I think, using the new features of pyo3 for struct variant enums:

```rust
#[pyclass]
enum PyStringFunction {
   ConcatHorizontal { delimiter: Whatever, ignore_nulls: bool },
   ...
}

...
match strfn {
   StringFunction::ConcatHorizontal { d, i } => PyStringFunction::ConcatHorizontal {d, i},
   ...
}
```

AFAICT, from some (very minimal) microbenchmarking, this all have pretty much the same performance characteristics.

One niggle for the last option is that (for now) you can't mix unit an struct enum variants, and tuple enum variants aren't supported at all, so you can't write:

```
#[pyclass]
enum SomeEnum {
   Unit,
   Tuple(i32),
   Struct { a: i32, b: i64 },
}
```

Instead one needs to map every to (possibly empty) structs and come up with a name for the unnamed tuple variant field.

Thoughts?

Reply:
I do like that more as we can more easily change our internals whilst maintaining the same exposed enums. It is some extra typing atm, but it gives us an extra indirection that will likely make it more maintainable.